File: sdk/python/zoo_sdk/agent.py
import base64
import time
from dataclasses import dataclass, field

import anthropic

from zoo_sdk import Sandbox

WIDTH, HEIGHT = 1280, 720


@dataclass
class Agent:
    sandbox: Sandbox
    model: str = "claude-sonnet-5"
    tool_version: str = "computer_20251124"
    beta: str = "computer-use-2025-11-24"
    max_steps: int = 50
    max_tokens: int = 4096
    system: str = "You control a Linux XFCE desktop. Take a screenshot first, act step by step, verify with screenshots."
    client: anthropic.Anthropic = field(default_factory=anthropic.Anthropic)
    on_step: object = None

    def tools(self) -> list[dict]:
        return [
            {"type": self.tool_version, "name": "computer", "display_width_px": WIDTH, "display_height_px": HEIGHT},
            {"type": "bash_20250124", "name": "bash"},
        ]

    def screenshot(self) -> dict:
        data = base64.b64encode(self.sandbox.screenshot()).decode()
        return {"type": "image", "source": {"type": "base64", "media_type": "image/png", "data": data}}

    def xdotool(self, args: str):
        self.sandbox.exec(f"DISPLAY=:1 xdotool {args}")

    def computer(self, action: str, coordinate=None, text=None, **kw) -> list[dict]:
        sb = self.sandbox
        x, y = coordinate or (None, None)
        if action in ("left_click", "right_click", "middle_click"):
            if text:
                self.xdotool(f"keydown {text}")
            sb.click(x=x, y=y, button=action.split("_")[0])
            if text:
                self.xdotool(f"keyup {text}")
        elif action == "double_click":
            sb.double_click(x=x, y=y)
        elif action == "triple_click":
            self.xdotool(f"mousemove {x} {y} click --repeat 3 1")
        elif action == "mouse_move":
            self.xdotool(f"mousemove {x} {y}")
        elif action == "left_click_drag":
            sx, sy = kw["start_coordinate"]
            sb.drag(start_x=sx, start_y=sy, end_x=x, end_y=y)
        elif action == "left_mouse_down":
            self.xdotool("mousedown 1")
        elif action == "left_mouse_up":
            self.xdotool("mouseup 1")
        elif action == "type":
            sb.type_text(text=text)
        elif action == "key":
            sb.press_key(key=text)
        elif action == "hold_key":
            self.xdotool(f"keydown {text} sleep {kw.get('duration', 1)} keyup {text}")
        elif action == "scroll":
            sb.scroll(direction=kw["scroll_direction"], amount=kw.get("scroll_amount", 3), x=x, y=y)
        elif action == "wait":
            time.sleep(kw.get("duration", 1))
        elif action == "cursor_position":
            out = sb.exec("DISPLAY=:1 xdotool getmouselocation")["stdout"]
            return [{"type": "text", "text": out}]
        elif action != "screenshot" and action != "zoom":
            return [{"type": "text", "text": f"unsupported action {action}"}]
        time.sleep(0.3)
        return [self.screenshot()]

    def bash(self, command: str | None = None, restart: bool = False, **_) -> list[dict]:
        if restart or not command:
            return [{"type": "text", "text": "ok"}]
        r = self.sandbox.exec(command, timeout=120)
        return [{"type": "text", "text": f"exit {r['exit_code']}\n{r['stdout']}\n{r['stderr']}"[-20000:]}]

    def run(self, task: str) -> str:
        messages = [{"role": "user", "content": task}]
        for _ in range(self.max_steps):
            response = self.client.beta.messages.create(
                model=self.model,
                max_tokens=self.max_tokens,
                system=self.system,
                tools=self.tools(),
                messages=messages,
                betas=[self.beta],
            )
            messages.append({"role": "assistant", "content": response.content})
            if self.on_step:
                self.on_step(response)
            calls = [b for b in response.content if b.type == "tool_use"]
            if not calls:
                return "".join(b.text for b in response.content if b.type == "text")
            results = []
            for call in calls:
                try:
                    content, error = getattr(self, call.name)(**call.input), False
                except Exception as e:
                    content, error = [{"type": "text", "text": str(e)}], True
                results.append({"type": "tool_result", "tool_use_id": call.id, "content": content, "is_error": error})
            messages.append({"role": "user", "content": results})
        raise RuntimeError(f"agent did not finish in {self.max_steps} steps")


def run(sandbox: Sandbox, task: str, **kwargs) -> str:
    return Agent(sandbox, **kwargs).run(task)

File: examples/cua-ts/agent.ts
import Anthropic from "@anthropic-ai/sdk"

const ZOO_URL = process.env.ZOO_URL ?? "http://localhost:8000"
const ZOO_API_KEY = process.env.ZOO_API_KEY!
const MODEL = process.env.MODEL ?? "claude-sonnet-5"
const TOOL_VERSION = process.env.TOOL_VERSION ?? "computer_20251124"
const BETA = process.env.BETA ?? "computer-use-2025-11-24"
const [WIDTH, HEIGHT] = [1280, 720]

type Content = Anthropic.Beta.BetaToolResultBlockParam["content"]

async function zoo(path: string, body?: unknown, raw = false) {
  const res = await fetch(`${ZOO_URL}${path}`, {
    method: body === undefined ? "GET" : "POST",
    headers: {
      Authorization: `Bearer ${ZOO_API_KEY}`,
      "Content-Type": "application/json",
    },
    body: body === undefined ? undefined : JSON.stringify(body),
  })
  if (!res.ok) throw new Error(`${res.status} ${await res.text()}`)
  return raw ? res : res.json()
}

async function createSandbox(name: string) {
  const sandbox = await zoo("/sandboxes", { name, kind: "desktop" })
  for (;;) {
    const s = await zoo(`/sandboxes/${sandbox.id}`)
    if (s.status === "running") return s.id as string
    if (s.status !== "provisioning") throw new Error(`sandbox ${s.status}: ${s.error_message}`)
    await Bun.sleep(1000)
  }
}

function tool(id: string, name: string, args: Record<string, unknown> = {}) {
  return zoo(`/sandboxes/${id}/tools/${name}`, args)
}

async function screenshot(id: string): Promise<Content> {
  const res = (await zoo(`/sandboxes/${id}/screenshot`, {}, true)) as Response
  const data = Buffer.from(await res.arrayBuffer()).toString("base64")
  return [{ type: "image", source: { type: "base64", media_type: "image/png", data } }]
}

const xdotool = (id: string, args: string) =>
  tool(id, "execute_command", { command: `DISPLAY=:1 xdotool ${args}` })

async function computer(id: string, input: Record<string, any>): Promise<Content> {
  const [x, y] = input.coordinate ?? []
  switch (input.action) {
    case "left_click":
    case "right_click":
    case "middle_click":
      await tool(id, "click", { x, y, button: input.action.split("_")[0] })
      break
    case "double_click":
      await tool(id, "double_click", { x, y })
      break
    case "triple_click":
      await xdotool(id, `mousemove ${x} ${y} click --repeat 3 1`)
      break
    case "mouse_move":
      await xdotool(id, `mousemove ${x} ${y}`)
      break
    case "left_click_drag": {
      const [sx, sy] = input.start_coordinate
      await tool(id, "drag", { start_x: sx, start_y: sy, end_x: x, end_y: y })
      break
    }
    case "type":
      await tool(id, "type_text", { text: input.text })
      break
    case "key":
      await tool(id, "press_key", { key: input.text })
      break
    case "scroll":
      await tool(id, "scroll", {
        direction: input.scroll_direction,
        amount: input.scroll_amount ?? 3,
        x,
        y,
      })
      break
    case "wait":
      await Bun.sleep((input.duration ?? 1) * 1000)
      break
    case "screenshot":
      break
    default:
      return [{ type: "text", text: `unsupported action ${input.action}` }]
  }
  await Bun.sleep(300)
  return screenshot(id)
}

async function bash(id: string, input: Record<string, any>): Promise<Content> {
  if (!input.command) return [{ type: "text", text: "ok" }]
  const r = await tool(id, "execute_command", { command: input.command, timeout: 120 })
  return [{ type: "text", text: `exit ${r.exit_code}\n${r.stdout}\n${r.stderr}`.slice(-20000) }]
}

async function run(id: string, task: string) {
  const client = new Anthropic()
  const messages: Anthropic.Beta.BetaMessageParam[] = [{ role: "user", content: task }]
  for (let step = 0; step < 50; step++) {
    const response = await client.beta.messages.create({
      model: MODEL,
      max_tokens: 4096,
      betas: [BETA],
      system: "You control a Linux XFCE desktop. Take a screenshot first, act step by step, verify with screenshots.",
      tools: [
        { type: TOOL_VERSION, name: "computer", display_width_px: WIDTH, display_height_px: HEIGHT } as any,
        { type: "bash_20250124", name: "bash" },
      ],
      messages,
    })
    messages.push({ role: "assistant", content: response.content })
    const calls = response.content.filter((b) => b.type === "tool_use")
    for (const b of response.content) if (b.type === "text") console.log(b.text)
    if (calls.length === 0) return
    const results: Anthropic.Beta.BetaToolResultBlockParam[] = []
    for (const call of calls) {
      console.log(`→ ${call.name}`, JSON.stringify(call.input))
      try {
        const handler = call.name === "computer" ? computer : bash
        results.push({ type: "tool_result", tool_use_id: call.id, content: await handler(id, call.input as any) })
      } catch (e) {
        results.push({ type: "tool_result", tool_use_id: call.id, content: String(e), is_error: true })
      }
    }
    messages.push({ role: "user", content: results })
  }
  throw new Error("agent did not finish in 50 steps")
}

const task = process.argv.slice(2).join(" ") || "Open Firefox and search for the weather in Paris"
const id = process.env.SANDBOX_ID ?? (await createSandbox("cua-ts"))
console.log(`sandbox ${id}`)
await run(id, task)
